auth/couchdb_session_authenticator.go

```diff
@@ -155,49 +160,59 @@ func (a *CouchDbSessionAuthenticator) Authenticate(request *http.Request) error
 // getCookie returns an AuthSession cookie to be used in a request.
 // A new cookie will be fetched from the session end-point when needed.
 func (a *CouchDbSessionAuthenticator) getCookie() (*http.Cookie, error) {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+
+	a.flushRefreshChannel()
+
 	if a.session == nil || !a.session.isValid() {
-		err := a.syncRequestSession()
+		newSession, err := a.requestSession()
 		if err != nil {
 			return nil, err
 		}
+		a.session = newSession
 	} else if a.session.needsRefresh() {
-		ch := make(chan error)
+		// start a background process to refresh session cookie
+		// if session returned fast enough it'll be updated at the select below
+		// otherwise it'll be updated in next request at flushRefreshChannel
+		ch := make(chan error, 1)
 		go func() {
-			ch <- a.requestSession()
+			session, err := a.requestSession()
+			if err != nil {
+				ch <- err
+				return
+			}
+			a.refresh <- session
 		}()
 		select {
```

Comment:
Well, you wrong. This select always reached, if there is anything on `ch` or `refresh` channel it processed, otherwise select does no-op on `default` and function carries on.